examples/BERT/ns_task.py

```diff
@@ -0,0 +1,248 @@
+        for idx, (seq_input, tok_type, target_ns_labels) in enumerate(dataloader):
+            if args.parallel == 'DDP':
+                seq_input = seq_input.to(device[0])
+                tok_type = tok_type.to(device[0])
+                target_ns_labels = target_ns_labels.to(device[0])
+            else:
+                seq_input = seq_input.to(device)
+                tok_type = tok_type.to(device)
+                target_ns_labels = target_ns_labels.to(device)
+            seq_input = seq_input.transpose(0, 1)  # Wrap up by DDP or DataParallel
+            ns_labels = model(seq_input, token_type_input=tok_type)
+            loss = criterion(ns_labels, target_ns_labels)
+            total_loss += loss.item()
+    return total_loss / (len(data_source) // batch_size)
+
+
+def train(train_dataset, model, train_loss_log, device, optimizer, criterion,
+          epoch, scheduler, cls_id, sep_id, pad_id, args, rank=None):
+    model.train()
+    total_loss = 0.
+    start_time = time.time()
+    batch_size = args.batch_size
+    dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True,
+                            collate_fn=lambda b: collate_batch(b, args, cls_id, sep_id, pad_id))
+    train_loss_log.append(0.0)
+    for idx, (seq_input, tok_type, target_ns_labels) in enumerate(dataloader):
+        if args.parallel == 'DDP':
+            seq_input = seq_input.to(device[0])
+            tok_type = tok_type.to(device[0])
+            target_ns_labels = target_ns_labels.to(device[0])
+        else:
+            seq_input = seq_input.to(device)
+            tok_type = tok_type.to(device)
+            target_ns_labels = target_ns_labels.to(device)
+        optimizer.zero_grad()
+        seq_input = seq_input.transpose(0, 1)  # Wrap up by DDP or DataParallel
+        ns_labels = model(seq_input, token_type_input=tok_type)
+        loss = criterion(ns_labels, target_ns_labels)
+        loss.backward()
+        torch.nn.utils.clip_grad_norm_(model.parameters(), args.clip)
+        optimizer.step()
+        total_loss += loss.item()
+        if idx % args.log_interval == 0 and idx > 0:
+            cur_loss = total_loss / args.log_interval
+            elapsed = time.time() - start_time
+            if (rank is None) or rank == 0:
+                train_loss_log[-1] = cur_loss
+                print('| epoch {:3d} | {:5d}/{:5d} batches | lr {:05.5f} | '
+                      'ms/batch {:5.2f} | '
+                      'loss {:8.5f} | ppl {:5.2f}'.format(epoch, idx,
+                                                          len(train_dataset) // batch_size,
+                                                          scheduler.get_last_lr()[0],
+                                                          elapsed * 1000 / args.log_interval,
+                                                          cur_loss, math.exp(cur_loss)))
+            total_loss = 0
+            start_time = time.time()
+
+
+def run_main(args, rank=None):
```

Comment:
It seems like some of this code used here could be shared with the other run_main functions?

Reply:
You are right. However, since those three tasks are quite different, I prefer to explicitly set up the `run_main` func